src/mpi/ccl/cclcomm.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#define CUDA_ERR_CHECK(ret) if (unlikely((ret) != cudaSuccess)) goto fn_fail
+
+/*
+ * CCLcomm functions, currently tied to NCCL
+ */
+
+int MPIR_CCLcomm_init(MPIR_Comm * comm, int rank)
+{
+    int mpi_errno = MPI_SUCCESS;
+    cudaError_t ret;
+    int comm_size = comm->local_size;
+
+    MPIR_CCLcomm *cclcomm;
+    cclcomm = MPL_malloc(sizeof(MPIR_CCLcomm), MPL_MEM_OTHER);
+    MPIR_ERR_CHKANDJUMP(!cclcomm, mpi_errno, MPI_ERR_OTHER, "**nomem");
+
+    if(rank == 0) {
+        ncclGetUniqueId(&(cclcomm->id));
+    }
+
+    mpi_errno = MPIR_Bcast_impl(&(cclcomm->id), sizeof(cclcomm->id), MPI_INT, 0, comm,
+                                    MPIR_ERR_NONE);
+    MPIR_ERR_CHECK(mpi_errno);
+    
+    ret = cudaSetDevice(0); //TODO: Fix this for multi-GPU nodes
+    CUDA_ERR_CHECK(ret);
+    ret = cudaStreamCreate(&(cclcomm->stream));
+    CUDA_ERR_CHECK(ret);
+    ret = ncclCommInitRank(&(cclcomm->ncclcomm), comm_size, cclcomm->id, rank);
+    CUDA_ERR_CHECK(ret);
+
+    cclcomm->comm = comm;
+    comm->cclcomm = cclcomm;
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIR_CCLcomm_free(MPIR_Comm * comm)
+{
+    int mpi_errno = MPI_SUCCESS;
+    cudaError_t ret;
+
+    MPIR_Assert(comm->cclcomm);
+    MPIR_CCLcomm *cclcomm = comm->cclcomm;
+
+    ret = cudaStreamSynchronize(cclcomm->stream);
+    CUDA_ERR_CHECK(ret);
+    ret = ncclCommDestroy(cclcomm->ncclcomm);
+    CUDA_ERR_CHECK(ret);
+    ret = cudaStreamDestroy(cclcomm->stream);
+    CUDA_ERR_CHECK(ret);
+
+    MPL_free(cclcomm);
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/*
+ * NCCL-specific functions
+ */
+
+int MPIR_NCCL_red_op_is_supported(MPI_Op op)
+{
+    switch (op) {
+      case MPI_SUM:
+      case MPI_PROD:
+      case MPI_MIN:
+      case MPI_MAX:
+        return 1;
+      default:
+        return 0;
+    }
+}
+
+int MPIR_NCCL_get_red_op(MPI_Op op, ncclRedOp_t *redOp)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    switch (op) {
+      case MPI_SUM:
+        *redOp = ncclSum;
+        break;
+      case MPI_PROD:
+        *redOp = ncclProd;
+        break;
+      case MPI_MIN:
+        *redOp = ncclMin;
+        break;
+      case MPI_MAX:
+        *redOp = ncclMax;
+        break;
+      default:
+        goto fn_fail;
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    mpi_errno = MPI_ERR_ARG;
+    goto fn_exit;
+}
+
+int MPIR_NCCL_datatype_is_supported(MPI_Datatype dtype)
```

Comment:
rebasing on #7264 is probably a good idea even before its merged. it should make some of this type stuff less verbose, I think.